eth/client.go

```diff
@@ -150,306 +226,601 @@ func NewTransactOptsForAccount(account accounts.Account, passphrase string, keyS
+// TRANSACTIONS
 
-	if err != nil {
-		glog.Errorf("Error decoding ABI into JSON: %v", err)
-		return nil, err
-	}
+func (c *Client) InitializeRound() (<-chan types.Receipt, <-chan error) {
+	outRes := make(chan types.Receipt)
+	outErr := make(chan error)
 
-	q := ethereum.FilterQuery{
-		Addresses: []common.Address{c.protocolAddr},
-		Topics:    [][]common.Hash{[]common.Hash{protocolJson.Events["NewJob"].Id()}, []common.Hash{common.BytesToHash(common.LeftPadBytes(c.account.Address[:], 32))}},
-	}
+	go func() {
+		defer close(outRes)
+		defer close(outErr)
 
-	return c.backend.SubscribeFilterLogs(ctx, q, logsCh)
+		tx, err := c.roundsManagerSession.InitializeRound()
+		if err != nil {
+			outErr <- err
+			return
+		}
+
+		glog.Infof("[%v] Submitted tx %v. Initialize round", c.account.Address.Hex(), tx.Hash().Hex())
+
+		receipt, err := c.WaitForReceipt(tx)
+		if err != nil {
+			outErr <- err
+		} else {
+			outRes <- *receipt
+		}
+
+		return
+	}()
+
+	return outRes, outErr
 }
 
-func monitorJobEvent(logsCh <-chan types.Log, callback func(l types.Log) error) {
-	for {
-		select {
-		case log := <-logsCh:
-			if !log.Removed && log.BlockNumber != 0 {
-				// TODO: Perform action in response to job event
-				// glog.Infof("Received a job at block %v with tx %v by address %v", log.BlockNumber, log.TxHash.Hex(), log.Address.Hex())
-				callback(log)
-			}
+func (c *Client) Transcoder(blockRewardCut uint8, feeShare uint8, pricePerSegment *big.Int) (<-chan types.Receipt, <-chan error) {
+	outRes := make(chan types.Receipt)
+	outErr := make(chan error)
+
+	go func() {
+		defer close(outRes)
+		defer close(outErr)
+
+		tx, err := c.bondingManagerSession.Transcoder(blockRewardCut, feeShare, pricePerSegment)
+		if err != nil {
+			outErr <- err
+			return
 		}
-	}
+
+		glog.Infof("[%v] Submitted tx %v. Register as transcoder", c.account.Address.Hex(), tx.Hash().Hex())
+
+		receipt, err := c.WaitForReceipt(tx)
+		if err != nil {
+			outErr <- err
+			return
+		}
+
+		outRes <- *receipt
+
+		return
+	}()
+
+	return outRes, outErr
 }
 
-func (c *Client) WatchEvent(logsCh <-chan types.Log) (types.Log, error) {
-	var (
-		timer = time.NewTimer(c.eventTimeout)
-	)
+func (c *Client) Bond(amount *big.Int, toAddr common.Address) (<-chan types.Receipt, <-chan error) {
+	inRes, inErr := c.Approve(c.bondingManagerAddr, amount)
+
+	timer := time.NewTimer(c.eventTimeout)
+	outRes := make(chan types.Receipt)
+	outErr := make(chan error)
+
+	go func() {
+		defer close(outRes)
+		defer close(outErr)
 
-	for {
 		select {
-		case log := <-logsCh:
+		case log := <-inRes:
 			if !log.Removed {
-				return log, nil
+				tx, err := c.bondingManagerSession.Bond(amount, toAddr)
+				if err != nil {
+					outErr <- err
+					return
+				}
+
+				glog.Infof("[%v] Submitted tx %v. Bond %v LPTU to %v", c.account.Address.Hex(), tx.Hash().Hex(), amount, toAddr.Hex())
+
+				receipt, err := c.WaitForReceipt(tx)
+				if err != nil {
+					outErr <- err
```

Comment:
Hm an additional transaction to reset the approval amount to 0 if the `bond` transaction fails does not seem great. Ideally, `bond` and `approve` can be bundled into an atomic transaction so that if any of the intermediate steps fail, the entire transaction is reverted. One way to accomplish this is to implement an `approveAndCall` function in `LivepeerToken` that approves a token amount and immediately triggers another function call in the same transaction. This is probably a contract design decision we can put off for now though and we can roll with an additional transaction for setting the approval amount to 0 for the time being